keps/sig-api-machinery/20190621-apply-scalability.md

```diff
@@ -0,0 +1,154 @@
+---
+title: KEP Template
+authors:
+  - "@jennybuckley"
+  - "@apelisse"
+owning-sig: sig-api-machinery
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-06-21
+last-updated: 2019-06-21
+status: provisional
+see-also:
+  - "/keps/sig-api-machinery/0006-apply.md"
+---
+
+# Apply Scalability
+
+## Table of Contents
+
+- TODO
+
+## Summary
+
+Serverside Apply (SSA) in its current implementation causes some serious scalability issues when the feature is enabled. Specifically, list operation duration was highly dependent on the number of updates happening. From our testing it seemed to be because of a large number of small object allocations.
+
+## Motivation
+
+### The Problems
+
+From our testing, the problems seem to be because of a large number of small object allocations.
+
+#### On list, update, and apply: 
+
+Deserializing managedFields as a deeply nested field in the apiserver allocates many small objects. Objects are deserialized when fetched from etcd.
+The size of managedFields is similar to the size of the object. It increases etcd size, and cache size of the apiserver and clients.
+
+#### On update and apply: 
+
+Converting managedFields from the api representation into the internal representation causes many small object allocations.
+Some structured-merge-diff operations (like compare) also create many small objects that are not needed. Most “field set” operations are not done in place.
+
+### Experiments
+
+The following experiments convinced us that the problem is mostly the allocations of small objects. Latency in etcd because of read/write with bigger objects, and/or data transfer was not a significant issue.
+
+For all these scenarios, we’ve created a deployment with 5,000 pods. Pods that have the managedFields (when server-side apply is enabled) are obviously larger. Here are some numbers for the comparison, with the objects in protobuf format:
+
+- Server-side apply disabled:
+  - A pod is about 8,137 bytes
+  - 5,000 pods is 40,506,271 bytes
+- Server-side apply enabled:
+  - A pod is about 13,164 bytes (~62% larger)
+  - 5,000 pods is 67,182,690 bytes
+
+All the following scenarios have been tested by LISTing all the pods, then running 250 patches per second while still LISTing the pods at the same time.
+
+1. Server-side apply disabled, 
+2. Server-side apply enabled, 
+3. Server-side apply disabled, but with an extra annotation. The annotation is about 5,000 bytes to compensate for the difference in object size with server-side apply.
+4. Server-side apply disabled, with 320 small annotations. Those annotations total about 5,000 bytes to simulate the small object allocation.
+
+The average latency are as follows:
+
+| Scenario | LIST | PATCH | LIST while PATCH |
+|---|---|---|---|
+| 1. |360.4 ms | 8.0 ms | 361.0 ms |
+| 2. | 1294.5 ms | 38.5 ms | 2529.6 ms |
```

Comment:
This is interesting. "LIST while PATCH" shows a marked difference between this and case 4.  Did you investigate the reason or will that come separately?

Reply:
Both LIST and PATCH do perform lots of allocations: LIST because of deserialization, PATCH because of deserialization and conversion from external/internal/external type

Reply:
> Is the managedField conceptually map[string]string or map[string]interface{} with some multilayer deserialization? - that's one thing that can affect it.

Yes, it is `map[string]interface{}` with multilayer deserialization

> Was your master machine (apiserver) cpu-saturated in these experiments?

I haven't checked that ..

Reply:
We need to update this with the latest numbers.